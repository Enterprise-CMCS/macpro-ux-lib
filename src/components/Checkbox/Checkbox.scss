@use "../../sass/macbis-colors" as c;

.usa-checkbox__input--tile+[class*=__label] {
  border: 1px solid c.$gray-lighter;
  border-radius: 0.2rem;
}

.usa-checkbox__input:checked+[class*=__label]::before {
  background-color: c.$primary;
  box-shadow: 0 0 0 2px c.$primary;
}

.usa-checkbox__input--tile:checked+[class*=__label] {
  background-color: c.$primary-alt-lightest;
  border-color: c.$primary;
}

.usa-checkbox__input--tile:disabled+[class*=__label] {
  background-color: c.$gray-lightest;
  border-color: c.$gray-lighter;
}

.usa-checkbox__input:disabled+[class*=__label]::before, .usa-radio__input[aria-disabled=true]+[class*=__label]::before {
  background-color: c.$gray-lighter;
  box-shadow: 0 0 0 2px rgb(27 27 27 / 30%);
}
