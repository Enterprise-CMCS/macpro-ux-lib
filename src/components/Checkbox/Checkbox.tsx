import React, { useState } from "react";

type IntrinsicElements = JSX.IntrinsicElements["input"];

interface Props extends IntrinsicElements {
  checked?: boolean;
  children?: JSX.Element[];
  id: string;
  isTile?: boolean;
  label: string;
  name: string;
  tileDescription?: string;
  value?: string;
}

/**
 * Checkbox Component
 * @param {string}  id                A unique identifier for the input.
 * @param {string}  label             Label text that appears to the right of the Checkbox.
 * @param {string}  name              Name of the input element.
 * @param {boolean} [checked]         Set the initial checked state.
 * @param {Array}   [children]        An array of child elements to appear when checked.
 * @param {boolean} [isTile]          Use the tile variation of the Checkbox.
 * @param {string}  [tileDescription] Text that can be used to describe the label in more detail. Activates the tile variation automatically.
 * @param {string}  [value]           Value of the input element.
 */
export const Checkbox: React.FC<Props> = ({
  checked = false,
  children,
  id,
  isTile = false,
  label,
  name,
  tileDescription,
  value,
  ...rest
}) => {
  const [isChecked, setChecked] = useState(checked);
  const handleChange = () => {
    setChecked(!isChecked);
  };

  return (
    <div className="usa-checkbox">
      <input
        checked={isChecked}
        className={
          isTile
            ? "usa-checkbox__input usa-checkbox__input--tile"
            : "usa-checkbox__input"
        }
        id={id}
        onChange={handleChange}
        type="checkbox"
        name={name}
        value={value}
        {...rest}
      />
      <label className="usa-checkbox__label" htmlFor={id}>
        {label}
        {isTile && tileDescription && (
          <span className="usa-checkbox__label-description">
            {tileDescription}
          </span>
        )}
      </label>
      {isChecked && children?.length && !isTile && (
        <div className="border-left-05 border-primary margin-left-1 padding-left-205">
          {children}
        </div>
      )}
    </div>
  );
};
