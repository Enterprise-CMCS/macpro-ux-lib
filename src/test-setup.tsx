import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
