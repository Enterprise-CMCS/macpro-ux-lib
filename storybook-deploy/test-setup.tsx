import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
