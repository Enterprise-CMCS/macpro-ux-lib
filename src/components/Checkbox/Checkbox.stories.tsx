import { ComponentMeta, ComponentStory } from "@storybook/react";
import React from "react";
import { Checkbox } from "./Checkbox";
import { TextArea } from "../TextArea/TextArea";

export default {
  title: "USWDS/Base/Checkbox",
  component: Checkbox,
  argTypes: {
    id: {
      description: "A unique identifier for the input.",
    },
    isTile: {
      description: "Use the tile variation of the Checkbox.",
    },
    label: {
      description: "Label text that appears to the right of the Checkbox.",
    },
    name: {
      description: "Name of the input element.",
    },
    tileDescription: {
      description:
        "Text that can be used to describe the label in more detail. Activates the tile variation automatically.",
    },
    value: {
      description: "Value of the input element.",
    },
  },
  args: {
    checked: false,
    disabled: false,
    isTile: false,
    tileDescription: "",
  },
} as ComponentMeta<typeof Checkbox>;

const Template: ComponentStory<typeof Checkbox> = ({ ...rest }) => (
  <Checkbox {...rest} />
);

export const Default = Template.bind({});
Default.args = {
  id: "checkbox-item-default",
  label: "Checkbox Item",
  name: "checkbox",
  value: "checkbox-item",
};

export const Disabled = Template.bind({});
Disabled.args = {
  id: "checkbox-item-disabled",
  label: "Checkbox Item",
  name: "checkbox",
  value: "checkbox-item",
  disabled: true,
};

export const Child = Template.bind({});
Child.args = {
  checked: true,
  id: "checkbox-item-child",
  label: "Checkbox Item",
  name: "checkbox",
  value: "checkbox-item",
  children: [<TextArea label="Text input label" fieldName="text-input" />],
};

export const MultipleChildren = Template.bind({});
MultipleChildren.args = {
  checked: true,
  id: "checkbox-item",
  label: "Checkbox Item",
  name: "checkbox",
  value: "checkbox-item",
  children: [
    <Checkbox
      checked={true}
      id="checkbox-item-child-1"
      label="Child 1"
      name="checkbox-children"
      value="child-1"
      children={[
        <TextArea label="Child 1's TextArea" fieldName="child-1-textArea" />,
      ]}
    />,
    <Checkbox
      checked={true}
      id="checkbox-item-child-2"
      label="Child 2"
      name="checkbox-children"
      value="child-2"
      children={[
        <TextArea label="Child 2's TextArea" fieldName="child-2-textArea" />,
        <Checkbox
          checked={true}
          id="checkbox-item-child-3"
          label="Child 2's Child"
          name="second-child"
          value="child-2-child"
          children={[
            <TextArea
              label="Child 2's Child TextArea"
              fieldName="child-2-child-textArea"
            />,
          ]}
        />,
      ]}
    />,
  ],
};

export const Tile = Template.bind({});
Tile.args = {
  id: "check-tile",
  isTile: true,
  label: "Checkbox Item",
  name: "check-tile-0",
  value: "tile",
};

export const InactiveTitle = Template.bind({});
InactiveTitle.args = {
  disabled: true,
  id: "check-tile",
  isTile: true,
  label: "Checkbox Item",
  name: "check-tile-0",
  value: "tile",
};

export const TileWithDescription = Template.bind({});
TileWithDescription.args = {
  id: "check-tile-description",
  isTile: true,
  label: "Checkbox Item",
  name: "check-tile-desc-0",
  value: "tile-children",
  tileDescription:
    "This is optional text that can be used to describe the label in more detail.",
};

export const InactiveTileWithDescription = Template.bind({});
InactiveTileWithDescription.args = {
  disabled: true,
  id: "check-tile-description",
  isTile: true,
  label: "Checkbox Item",
  name: "check-tile-desc-0",
  value: "tile-children",
  tileDescription:
    "This is optional text that can be used to describe the label in more detail.",
};
