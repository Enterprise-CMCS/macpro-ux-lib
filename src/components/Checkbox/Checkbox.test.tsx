import { Checkbox } from "./Checkbox";
import { render, screen } from "../../test-setup";
import fireEvent from "@testing-library/user-event";
import React from "react";

describe("Checkbox", () => {
  describe("Default", () => {
    beforeEach(() => {
      render(
        <form data-testid="test-form">
          <Checkbox
            id="check-default"
            label="Default Checkbox"
            name="test-checkbox"
            value="default-checkbox"
          />
        </form>
      );
    });

    it("is unchecked by default", () => {
      const cb = screen.getByLabelText("Default Checkbox");
      expect(cb).not.toBeChecked();
    });

    it("becomes checked on click", () => {
      const cb = screen.getByLabelText("Default Checkbox");
      fireEvent.click(cb);
      expect(cb).toBeChecked();
    });

    it("renders and has the expected label", () => {
      const cb = screen.getByLabelText("Default Checkbox");
      expect(cb).toBeInTheDocument();
    });

    it("is not a tile by default", () => {
      const cb = screen.getByLabelText("Default Checkbox");
      expect(cb).not.toHaveClass("usa-checkbox__input--tile");
    });

    it("has the expected ID", () => {
      const cb = screen.getByLabelText("Default Checkbox");
      expect(cb.getAttribute("id")).toBe("check-default");
    });

    it("has the expected name attribute", () => {
      const cb = screen.getByLabelText("Default Checkbox");
      expect(cb.getAttribute("name")).toBe("test-checkbox");
    });

    it("has the expected value when unchecked", () => {
      const testForm = screen.getByTestId("test-form");
      const cb = screen.getByLabelText("Default Checkbox");
      expect(testForm).toHaveFormValues({ "test-checkbox": false });
    });

    it("has the expected value when checked", () => {
      const testForm = screen.getByTestId("test-form");
      const cb = screen.getByLabelText("Default Checkbox");
      fireEvent.click(cb);
      expect(testForm).toHaveFormValues({ "test-checkbox": true });
    });
  });

  describe("CheckBox with children", () => {
    beforeEach(() => {
      render(
        <form data-testid="test-form">
          <Checkbox
            id="checkbox-with-children"
            label="Checkbox with Children"
            name="checkbox"
            value="with-children"
            children={[
              <Checkbox
                id="checkbox-item-child-1"
                label="Child 1"
                name="checkbox-children"
                value="child-1"
              />,
              <Checkbox
                id="checkbox-item-child-2"
                label="Child 2"
                name="checkbox-children"
                value="child-2"
              />,
            ]}
          />
        </form>
      );
    });

    it("shows the child element", () => {
      const cb = screen.getByLabelText("Checkbox with Children");
      fireEvent.click(cb);
      expect(cb).toBeChecked();

      const child = screen.getByLabelText("Child 1");
      expect(child).toBeInTheDocument();
    });

    it("has the expected form value when children are selected", () => {
      const testForm = screen.getByTestId("test-form");
      const cb = screen.getByLabelText("Checkbox with Children");
      fireEvent.click(cb);

      const child = screen.getByLabelText("Child 1");
      const child2 = screen.getByLabelText("Child 2");

      expect(testForm).toHaveFormValues({ checkbox: true });

      fireEvent.click(child);
      expect(child).toBeChecked();
      expect(testForm).toHaveFormValues({ "checkbox-children": ["child-1"] });

      fireEvent.click(child2);
      expect(child2).toBeChecked();
      expect(testForm).toHaveFormValues({
        "checkbox-children": ["child-1", "child-2"],
      });
    });
  });

  describe("Tile", () => {
    beforeEach(() => {
      render(
        <form data-testid="test-form">
          <Checkbox
            id="check-tile"
            isTile={true}
            label="Tile Checkbox"
            name="test-checkbox"
            tileDescription="This is the tile description."
            value="tile-checkbox"
          />
        </form>
      );
    });

    it("is renders as a tile", () => {
      const cb = screen.getByRole("checkbox");
      expect(cb).toHaveClass("usa-checkbox__input--tile");
    });

    it("is unchecked by default", () => {
      const cb = screen.getByRole("checkbox");
      expect(cb).not.toBeChecked();
    });

    it("becomes checked on click", () => {
      const cb = screen.getByRole("checkbox");
      fireEvent.click(cb);
      expect(cb).toBeChecked();
    });

    it("renders and has the expected label", () => {
      const cb = screen.getByRole("checkbox");
      expect(cb).toBeInTheDocument();
    });

    it("has the expected ID", () => {
      const cb = screen.getByRole("checkbox");
      expect(cb.getAttribute("id")).toBe("check-tile");
    });

    it("has the expected name attribute", () => {
      const cb = screen.getByRole("checkbox");
      expect(cb.getAttribute("name")).toBe("test-checkbox");
    });

    it("has the expected value when unchecked", () => {
      const testForm = screen.getByTestId("test-form");
      expect(testForm).toHaveFormValues({ "test-checkbox": false });
    });

    it("has the expected value when checked", () => {
      const testForm = screen.getByTestId("test-form");
      const cb = screen.getByRole("checkbox");
      fireEvent.click(cb);
      expect(testForm).toHaveFormValues({ "test-checkbox": true });
    });

    it("has the expected description text", () => {
      const testForm = screen.getByTestId("test-form");
      expect(testForm).toHaveTextContent("This is the tile description.");
    });
  });

  describe("Snapshot tests", () => {
    it("renders default Checkbox as expected", () => {
      const { container } = render(
        <Checkbox
          id="check-default"
          label="Default Checkbox"
          name="test-checkbox"
          value="default-checkbox"
        />
      );
      expect(container).toMatchSnapshot();
    });

    it("renders default Checkbox with children as expected", () => {
      const { container } = render(
        <Checkbox
          checked={true}
          id="checkbox-with-children"
          label="Checkbox with Children"
          name="checkbox"
          value="with-children"
          children={[
            <Checkbox
              id="checkbox-item-child-1"
              label="Child 1"
              name="checkbox-children"
              value="child-1"
            />,
          ]}
        />
      );
      expect(container).toMatchSnapshot();
    });

    it("renders tile Checkbox as expected", () => {
      const { container } = render(
        <Checkbox
          id="check-tile"
          isTile={true}
          label="Tile Checkbox"
          name="test-checkbox"
          tileDescription="This is the tile description."
          value="tile-checkbox"
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
