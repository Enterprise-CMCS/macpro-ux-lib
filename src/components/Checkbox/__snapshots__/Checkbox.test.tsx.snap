// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Snapshot tests renders default Checkbox as expected 1`] = `
<div>
  <div
    class="usa-checkbox"
  >
    <input
      class="usa-checkbox__input"
      id="check-default"
      name="test-checkbox"
      type="checkbox"
      value="default-checkbox"
    />
    <label
      class="usa-checkbox__label"
      for="check-default"
    >
      Default Checkbox
    </label>
  </div>
</div>
`;

exports[`Checkbox Snapshot tests renders default Checkbox with children as expected 1`] = `
<div>
  <div
    class="usa-checkbox"
  >
    <input
      checked=""
      class="usa-checkbox__input"
      id="checkbox-with-children"
      name="checkbox"
      type="checkbox"
      value="with-children"
    />
    <label
      class="usa-checkbox__label"
      for="checkbox-with-children"
    >
      Checkbox with Children
    </label>
    <div
      class="border-left-05 border-primary margin-left-1 padding-left-205"
    >
      <div
        class="usa-checkbox"
      >
        <input
          class="usa-checkbox__input"
          id="checkbox-item-child-1"
          name="checkbox-children"
          type="checkbox"
          value="child-1"
        />
        <label
          class="usa-checkbox__label"
          for="checkbox-item-child-1"
        >
          Child 1
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`Checkbox Snapshot tests renders tile Checkbox as expected 1`] = `
<div>
  <div
    class="usa-checkbox"
  >
    <input
      class="usa-checkbox__input usa-checkbox__input--tile"
      id="check-tile"
      name="test-checkbox"
      type="checkbox"
      value="tile-checkbox"
    />
    <label
      class="usa-checkbox__label"
      for="check-tile"
    >
      Tile Checkbox
      <span
        class="usa-checkbox__label-description"
      >
        This is the tile description.
      </span>
    </label>
  </div>
</div>
`;
